### PR TITLE
[NEW] Support for custom headers with providers

### PR DESCRIPTION
**Pull request by brad-larson**
This allows the user to supply arbitrary headers when constructing the
providers.

### CODE REVIEW

File: packages/rpc-provider/src/ws/Provider.ts [hunk 1]
```diff
@@ -138,7 +141,9 @@ export default class WsProvider implements WSProviderInterface {
 
       const WS = await getWSClass();
 
-      this.#websocket = new WS(this.#endpoints[this.#endpointIndex]);
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore - WS may be an instance of w3cwebsocket, which supports headers
```
Comment (brad-larson) [inline packages/rpc-provider/src/ws/Provider.ts:145]: There is likely a more elegant way to do this - but I wasn't sure which approach was preferred. Happy to update in that case

**Comment by jacogr** [comment]
Thank you, took a really quick couple of minutes glance and it seems fine, will do a proper look-through ASAP.

TIL ... standard (browser) WebSocket does not support custom headers. (In all honesty, did actually know about this, but it has been a while since actively woking with it so it has not been front-of mind.) So effectively it is only an option in the Node.js environment. Not against it at all in this (more-limited) case, it is just disappointing.

File: packages/rpc-provider/src/ws/Provider.ts (hunk 1 above)
Comment (jacogr) [reply packages/rpc-provider/src/ws/Provider.ts:145]: So this is where not doing this 100% in `getWSClass` bytes us. So generally, it would be -

```js
function getWSClass (): typeof WebSocket | typeof W3CWebSocket { ...} 

...

this.websocket = WS instance WebSocket
  ? new WS(...)
  : new WS(...)
```

Not the end of the world (although we know TS-overrides a a bit smelly), but the imp[act otherwise if probably much worse. So the comment here is probably the lesser of all evils and should rather be left as-is until the typing of `getWsClass` is done properly. (And not hacked about via `as unknown as ...` - which created the above)

TL;DR Certainly not perfect, but follows the lead of an issue introduced elsewhere - so with the comment, should be all good.

File: packages/rpc-provider/src/ws/Provider.ts [hunk 2]
```diff
@@ -138,7 +141,9 @@ export default class WsProvider implements WSProviderInterface {
...
+      this.#websocket = new WS(this.#endpoints[this.#endpointIndex], null, null, this.#headers);
```
Comment (jacogr) [inline packages/rpc-provider/src/ws/Provider.ts:146]:
```suggestion
      this.#websocket = new WS(this.#endpoints[this.#endpointIndex], undefined, undefined, this.#headers);
```

type definition is `constructor(
        url: string,
        protocols?: string | string[],
        origin?: string,
        headers?: http.OutgoingHttpHeaders,
        requestOptions?: object,
        IClientConfig?: IClientConfig,
    );`

Comment (jacogr) [reply packages/rpc-provider/src/ws/Provider.ts:146]: Would also suggest, looking at the above -

```js
import { isChildClass } from '@polkadot/util`;

...
this.#websocket = typeof WebSocket !== 'undefined' && isChildClass(WebSocket, WS)
  ? new WS(this.#endpoints[this.#endpointIndex])
  // // @ts-ignore
  : new WS(this.#endpoints[this.#endpointIndex], undefined, undefined, this.#headers);
```

this should (hopefully, maybe) fix the test failures as well?

Comment (brad-larson) [reply packages/rpc-provider/src/ws/Provider.ts]: Updated!

**Comment by jacogr** [comment]
Thank you very much, really appreciate it!

**Comment by iicc1** [comment]
Thanks for your contribution. I was wondering how to use the API with Auth Basic.
I think this should be documented in the docs, I spent hours trying to figure out how to correctly add the headers until I saw the test file of this pull.
In case someone is wondering, here it is how it is done:
`const provider = new WsProvider('ws/s://ip:port', retry, { Authorization: 'Basic xxxx' })`

**Comment by jacogr** [comment]
PRs to the docs are welcome and always appreciated.
